src/ConnectedMode.UnitTests/BranchMatcherTests.cs

```diff
@@ -0,0 +1,213 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using LibGit2Sharp;
+using SonarLint.VisualStudio.ConnectedMode.UnitTests.LibGit2SharpWrappers;
+using SonarQube.Client;
+using SonarQube.Client.Models;
+
+namespace SonarLint.VisualStudio.ConnectedMode.UnitTests
+{
+    [TestClass]
+    public class BranchMatcherTests
+    {
+        [TestMethod]
+        public async Task GetMatchedBranch_ChooseBranchWithSameName()
+        {
+            var service = CreateService("master", "branch");
+
+            var masterCommit = new CommitWrapper(1);
+            var branchCommit = new CommitWrapper(2);
+
+            var masterBranch = CreateBranch("master", masterCommit);
+            var headBranch = CreateBranch("branch", branchCommit, masterCommit);
+
+            var repo = CreateRepo(headBranch, masterBranch);
+
+            var testSubject = new BranchMatcher(service, repo);
+
+            var result = await testSubject.GetMatchedBranch("projectKey");
+
+            result.Should().Be("branch");
+        }
+
+        [TestMethod]
+        public async Task GetMatchedBranch_NoBranchWithSameName_ChooseClosestMatch()
+        {
+            var service = CreateService("master", "dev");
+
+            var masterCommit = new CommitWrapper(1);
+            var devBranchCommit = new CommitWrapper(2);
+            var headBranchCommit = new CommitWrapper(3);
+
+            var masterBranch = CreateBranch("master", masterCommit);
+            var devBranch = CreateBranch("dev", devBranchCommit, masterCommit);
+            var headBranch = CreateBranch("branch", headBranchCommit, devBranchCommit, masterCommit);
+
+            var repo = CreateRepo(headBranch, masterBranch, devBranch);
+
+            var testSubject = new BranchMatcher(service, repo);
+
+            var result = await testSubject.GetMatchedBranch("projectKey");
+
+            result.Should().Be("dev");
+        }
+
+        /*
```

Comment:
Added tree only for this test because other cases are very easy to understand by just looking at the code.